pkg/resource/deploy/plan_executor.go

```diff
@@ -184,29 +197,35 @@ func (pe *planExecutor) Execute(callerCtx context.Context, opts Options, preview
 
 // handleSingleEvent handles a single source event. For all incoming events, it produces a chain that needs
 // to be executed and schedules the chain for execution.
-func (pe *planExecutor) handleSingleEvent(event SourceEvent) error {
+func (pe *planExecutor) handleSingleEvent(event SourceEvent) (controlFlow, error) {
 	contract.Require(event != nil, "event != nil")
 
 	var steps []Step
+	var flow controlFlow
 	var err error
 	switch e := event.(type) {
 	case RegisterResourceEvent:
-		log.Infof("planExecutor.handleSingleEvent(...): received RegisterResourceEvent")
-		steps, err = pe.stepGen.GenerateSteps(e)
+		logging.V(4).Infof("planExecutor.handleSingleEvent(...): received RegisterResourceEvent")
+		steps, flow, err = pe.stepGen.GenerateSteps(e)
 	case ReadResourceEvent:
-		log.Infof("planExecutor.handleSingleEvent(...): received ReadResourceEvent")
-		steps, err = pe.stepGen.GenerateReadSteps(e)
+		logging.V(4).Infof("planExecutor.handleSingleEvent(...): received ReadResourceEvent")
+		steps, flow, err = pe.stepGen.GenerateReadSteps(e)
 	case RegisterResourceOutputsEvent:
-		log.Infof("planExecutor.handleSingleEvent(...): received register resource outputs")
+		logging.V(4).Infof("planExecutor.handleSingleEvent(...): received register resource outputs")
 		pe.stepExec.ExecuteRegisterResourceOutputs(e)
-		return nil
+		return Continue, nil
 	}
 
+	logging.V(4).Infof("planExecutor.handleSingleEvent(...): generated steps, err=%s, status=%d", err, flow)
 	if err != nil {
-		return err
+		return flow, err
```

Comment:
this is both interesting and hard to reason about.  We got an error from a deeper path, but we're also passing along flow control dcisions from a deeper level to a higher one.  it's not clear to me that that's necessarily the right thing to do.